### PR TITLE
Add Remote Auth IDP Page

### PR DESCRIPTION
**Pull request by jcaianirh**
Address Jira story: https://jira.coreos.com/browse/CONSOLE-1436

Add Remote Authentication IPD Page to Cluster Settings

Screen Shot 2019-05-01 at 5 30 42 PM

cc: @bmignano

### CODE REVIEW

**Comment by spadgett** [comment]
cc @enj

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+            The OAuth server base URL.
+          </p>
+        </div>
+        <IDPCAFileInput value={caFileContent} onChange={this.caFileChanged} />
+        <div className="form-group">
+          <DroppableFileInput
+            onChange={this.certFileChanged}
+            inputFileData={certFileContent}
+            id="cert-file-input"
+            label="certFile"
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]:
```suggestion
            label="Certificate"
```

We should add a help block:

> PEM-encoded TLS client certificate to present when connecting to the server.

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            hideContents />
+        </div>
+        <div className="form-group">
+          <DroppableFileInput
+            onChange={this.keyFileChanged}
+            inputFileData={keyFileContent}
+            id="key-file-input"
+            label="keyFile"
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]:
```suggestion
            label="Key"
```

with description

> PEM-encoded TLS private key for the client certificate. Required if certFile is specified.

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+    const secret = {
+      apiVersion: 'v1',
+      kind: 'Secret',
+      metadata: {
+        generateName: 'remoteauth-tlsclientcert-secret-',
+        namespace: 'openshift-config',
+      },
+      stringData: {
+        certFileContent,
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]:
```suggestion
        'tls.crt': certFileContent,
```

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+    const secret = {
+      apiVersion: 'v1',
+      kind: 'Secret',
+      metadata: {
+        generateName: 'remoteauth-tlsclientkey-secret-',
+        namespace: 'openshift-config',
+      },
+      stringData: {
+        keyFileContent,
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]:
```suggestion
        'tls.key': keyFileContent,
```

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+    if (_.isEmpty(this.state.keyFileContent) !== _.isEmpty(this.state.certFileContent)) {
+      this.setState({errorMessage: 'Values for certFile and keyFile should both be either excluded or provided.'});
+      return;
+    }
+    // Clear any previous errors.
+    this.setState({errorMessage: ''});
+    this.getOAuthResource().then((oauth: OAuthKind) => {
+      const promises = [
+        this.createCertSecret(),
+        this.createKeySecret(),
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]: I think you should be able to put both the cert and key in a single secret

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+import {ButtonBar, PromiseComponent, history, AsyncComponent} from '../utils';
+import { addIDP, getOAuthResource, redirectToOAuthPage } from './';
+import { IDPNameInput } from './idp-name-input';
+import { IDPCAFileInput } from './idp-cafile-input';
+
+const DroppableFileInput = (props: any) => <AsyncComponent loader={() => import('../utils/file-input').then(c => c.DroppableFileInput)} {...props} />;
+
+export class AddRemoteAuthPage extends PromiseComponent<{}, AddRemoteAuthPageState> {
+  readonly state: AddRemoteAuthPageState = {
+    name: 'remoteauth',
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]:
```suggestion
    name: 'basic-auth',
```

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+import { SecretModel, ConfigMapModel } from '../../models';
+import { IdentityProvider, k8sCreate, K8sResourceKind, OAuthKind } from '../../module/k8s';
+import {ButtonBar, PromiseComponent, history, AsyncComponent} from '../utils';
+import { addIDP, getOAuthResource, redirectToOAuthPage } from './';
+import { IDPNameInput } from './idp-name-input';
+import { IDPCAFileInput } from './idp-cafile-input';
+
+const DroppableFileInput = (props: any) => <AsyncComponent loader={() => import('../utils/file-input').then(c => c.DroppableFileInput)} {...props} />;
+
+export class AddRemoteAuthPage extends PromiseComponent<{}, AddRemoteAuthPageState> {
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]: s/RemoteAuth/BasicAuth throughout this file

```suggestion
export class AddBasicAuthPage extends PromiseComponent<{}, AddBasicAuthPageState> {
```

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    const { certFileContent } = this.state;
+    if (!certFileContent) {
+      return Promise.resolve(null);
+    }
+
+    const secret = {
+      apiVersion: 'v1',
+      kind: 'Secret',
+      metadata: {
+        generateName: 'remoteauth-tlsclientcert-secret-',
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]: Maybe something shorter like

```suggestion
        generateName: 'basic-auth-tls-',
```

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+    const { caFileContent } = this.state;
+    if (!caFileContent) {
+      return Promise.resolve(null);
+    }
+
+    const ca = {
+      apiVersion: 'v1',
+      kind: 'ConfigMap',
+      metadata: {
+        generateName: 'remoteauth-ca-',
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]:
```suggestion
        generateName: 'basic-auth-ca-',
```

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+    this.setState({certFileContent});
+  };
+
+  keyFileChanged = (keyFileContent: string) => {
+    this.setState({keyFileContent});
+  };
+
+  render() {
+    const { name, url, caFileContent, certFileContent, keyFileContent } = this.state;
+    const title = 'Add Identity Provider: Remote Authentication';
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]:
```suggestion
    const title = 'Add Identity Provider: Basic Authentication';
```

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+          <label className="control-label co-required" htmlFor="url">URL</label>
+          <input className="form-control"
+            type="text"
+            onChange={this.urlChanged}
+            value={url}
+            id="url"
+            aria-describedby="idp-url-help"
+            required />
+          <p className="help-block" id="idp-url-help">
+            The OAuth server base URL.
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]:
```suggestion
            The remote URL to connect to.
```

File: frontend/public/components/cluster-settings/oauth.tsx [hunk 12]
```diff
@@ -58,6 +58,7 @@ const OAuthDetails: React.SFC<OAuthDetailsProps> = ({obj}: {obj: OAuthKind}) =>
     htpasswd: 'HTPasswd',
     ldap: 'LDAP',
     oidconnect: 'OpenID Connect',
+    remoteauth: 'Remote Authentication',
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/oauth.tsx]:
```suggestion
    basicauth: 'Basic Authentication',
```

File: frontend/public/components/app.jsx [hunk 13]
```diff
@@ -268,6 +268,7 @@ class App extends React.PureComponent {
                   <LazyRoute path="/settings/idp/htpasswd" exact loader={() => import('./cluster-settings/htpasswd-idp-form' /* webpackChunkName: "htpasswd-idp-form" */).then(m => m.AddHTPasswdPage)} />
                   <LazyRoute path="/settings/idp/ldap" exact loader={() => import('./cluster-settings/ldap-idp-form' /* webpackChunkName: "ldap-idp-form" */).then(m => m.AddLDAPPage)} />
                   <LazyRoute path="/settings/idp/oidconnect" exact loader={() => import('./cluster-settings/openid-idp-form' /* webpackChunkName: "openid-idp-form" */).then(m => m.AddOpenIDPage)} />
+                  <LazyRoute path="/settings/idp/remoteauth" exact loader={() => import('./cluster-settings/remoteauth-idp-form' /* webpackChunkName: "remoteauth-idp-form" */).then(m => m.AddRemoteAuthPage)} />
```
Comment (spadgett) [inline frontend/public/components/app.jsx]:
```suggestion
                  <LazyRoute path="/settings/idp/basicauth" exact loader={() => import('./cluster-settings/basicauth-idp-form' /* webpackChunkName: "basicauth-idp-form" */).then(m => m.AddBasicAuthPage)} />
```

File: frontend/public/components/cluster-settings/remoteauth-idp-form.tsx (hunk 1 above)
Comment (jcaianirh) [reply frontend/public/components/cluster-settings/remoteauth-idp-form.tsx]: @spadgett I was using the designs here: https://docs.google.com/document/d/15MUwbd92RsZp5NMB1zyBzUwDBGupz90wqrku0iyC76c/edit#heading=h.arp1o27fsss9
I'll adjust based on your comments.

**Comment by jcaianirh** [comment]
@spadgett made requested code changes

File: frontend/public/components/cluster-settings/basicauth-idp-form.tsx [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+      };
+    }
+
+    return this.handlePromise(addIDP(oauth, idp));
+  }
+
+  submit: React.FormEventHandler<HTMLFormElement> = (e) => {
+    e.preventDefault();
+    if (_.isEmpty(this.state.keyFileContent) !== _.isEmpty(this.state.certFileContent)) {
+      this.setState({errorMessage: 'Values for certFile and keyFile should both be either excluded or provided.'});
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/basicauth-idp-form.tsx]:
```suggestion
      this.setState({errorMessage: 'Values for Certificate and Key should both be either excluded or provided.'});
```

File: frontend/public/components/cluster-settings/basicauth-idp-form.tsx [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+            inputFieldHelpText="PEM-encoded TLS private key for the client certificate. Required if certFile is specified." />
+        </div>
+        <div className="form-group">
+          <DroppableFileInput
+            onChange={this.keyFileChanged}
+            inputFileData={keyFileContent}
+            id="key-file-input"
+            label="Key"
+            hideContents
+            inputFieldHelpText="Required if certFile is specified." />
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/basicauth-idp-form.tsx]:
```suggestion
            inputFieldHelpText="PEM-encoded TLS private key for the client certificate. Required if Certificate is specified." />
```

File: frontend/public/components/cluster-settings/basicauth-idp-form.tsx [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+        </div>
+        <IDPCAFileInput value={caFileContent} onChange={this.caFileChanged} />
+        <div className="form-group">
+          <DroppableFileInput
+            onChange={this.certFileChanged}
+            inputFileData={certFileContent}
+            id="cert-file-input"
+            label="Certificate"
+            hideContents
+            inputFieldHelpText="PEM-encoded TLS private key for the client certificate. Required if certFile is specified." />
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/basicauth-idp-form.tsx]:
```suggestion
            inputFieldHelpText="PEM-encoded TLS client certificate to present when connecting to the server." />
```

File: frontend/public/components/cluster-settings/oauth.tsx [hunk 17]
```diff
@@ -58,6 +58,7 @@ const OAuthDetails: React.SFC<OAuthDetailsProps> = ({obj}: {obj: OAuthKind}) =>
     htpasswd: 'HTPasswd',
     ldap: 'LDAP',
     oidconnect: 'OpenID Connect',
+    basicauth: 'Basic Authentication',
```
Comment (spadgett) [inline frontend/public/components/cluster-settings/oauth.tsx]: I'd move this to the top to keep the list alphabetized

Comment (jcaianirh) [reply frontend/public/components/cluster-settings/oauth.tsx]: @spadgett good point it was alpha when it was Remote :)

**Comment by jcaianirh** [comment]
@spadgett thanks, fixed the wording as suggested.

**Pull request merged**